pkg/network/java/hotspot.go

```diff
@@ -0,0 +1,262 @@
+	dstPath = h.cwd + "/" + filepath.Base(agent)
+	if dst, err := os.Stat(h.root + dstPath); err == nil {
+		// if the destination file already exist
+		// check if it's not the source agent file
+		if src, err := os.Stat(agent); err == nil {
+			s := src.Sys().(*syscall.Stat_t)
+			d := dst.Sys().(*syscall.Stat_t)
+			if s.Dev == d.Dev && s.Ino == d.Ino {
+				return "", func() {}, nil
+			}
+		}
+	}
+
+	fagent, err := os.Open(agent)
+	if err != nil {
+		return "", func() {}, err
+	}
+	defer fagent.Close()
+
+	dst, err := os.OpenFile(h.root+dstPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, os.FileMode(0444))
+	if err != nil {
+		return "", func() {}, err
+	}
+	_, err = io.Copy(dst, fagent)
+	if err != nil {
+		return "", func() {}, err
+	}
+	dst.Close()
+	if err := syscall.Chown(h.root+dstPath, uid, gid); err != nil {
+		return "", func() {}, err
+	}
+
+	return dstPath, func() {
+		os.Remove(h.root + dstPath)
+	}, nil
+}
+
+func (h *Hotspot) connect() (cleanup func(), err error) {
+	JVMSocketPath := fmt.Sprintf("%s/.java_pid%d", h.tmpPath(), h.nsPid)
+	addr, err := net.ResolveUnixAddr("unix", JVMSocketPath)
+	if err != nil {
+		return func() {}, err
+	}
+	conn, err := net.DialUnix("unix", nil, addr)
+	if err != nil {
+		return func() {}, err
+	}
+
+	if err := conn.SetDeadline(time.Now().Add(3 * time.Second)); err != nil {
+		return func() {}, err
+	}
+	h.conn = conn
+	return func() { defer h.conn.Close() }, nil
+}
+
+func (h *Hotspot) command(cmd string, tailingNull bool) error {
+	if _, err := h.conn.Write([]byte{'1', 0}); err != nil { // Protocol version
+		return err
+	}
+
+	for _, c := range strings.Split(cmd, " ") {
+		cmd := append([]byte(c), 0)
+		if _, err := h.conn.Write(cmd); err != nil {
+			return err
+		}
+	}
+	if tailingNull {
+		if _, err := h.conn.Write([]byte{0}); err != nil {
+			return err
+		}
+	}
+
+	buf := make([]byte, 8192)
+	if _, err := h.conn.Read(buf); err != nil {
+		return err
+	}
+
+	returnCommand := 0
+	returnCode := 0
+	responseText := ""
+	line := 0
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		s := string(scanner.Bytes())
+		if line == 0 {
+			returnCommand, _ = strconv.Atoi(s)
+		}
+		if line == 1 && strings.HasPrefix(s, "return code: ") {
+			returnCode, _ = strconv.Atoi(s[len("return code: "):])
+		}
+		responseText += s + "\n"
+		line++
+	}
+	if returnCommand != 0 {
+		return fmt.Errorf("command '%s' return command %d code %d\n%s\n", cmd, returnCommand, returnCode, responseText)
+	}
+	return nil
+}
+
+// the (short) protocol is following
+//  o create a file .attach_pid%d
+//  o send a SIGQUIT signal
+//  o wait for socket file created by the java process
+func (h *Hotspot) attachJVMProtocol() error {
+	attachPath := fmt.Sprintf("%s/.attach_pid%d", h.root+h.cwd, h.nsPid)
+	hook, err := os.OpenFile(attachPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0660)
+	hook.Close()
+	hookUID, _, ownerErr := getPathOwner(attachPath)
+	if err != nil || ownerErr != nil || hookUID != uint32(os.Geteuid()) {
```

Comment:
why not trying to change the euid if we have root privileges in this context? (similar to what `jattach` do [here](https://github.com/apangin/jattach/blob/master/src/posix/jattach.c#L47-L50)